### ISSUE TITLE
Add classes also to Gothic 1

### ISSUE THREAD

**Issue by szapp** [Lehona/Ikarus#15]
Some classes provided for Gothic 2 are not available for Gothic 1. Most of them can probably just be copied from the Gothic 2 to the Gothic 1 scripts.

**Comment by szapp** [comment]
Here is a list in alphabetic order of class definitions provided for Gothic 1 and Gothic 2.

```diff
 Both
-Gothic 1 only
+Gothic 2 only
 

 CGameManager
 oCAniCtrl_Human
+oCBarrier
 oCGame
 oCInfo
 oCInfoChoice
 oCInfoManager
 oCInformationManager 
 oCItem
+oCLogManager
+oCLogTopic
 oCMag_Book
 oCMob
 oCMobContainer
 oCMobDoor
 oCMobFire
 oCMobInter
 oCMobLockable
-oCNews
 oCNpc
+oCNpcTalent
 oCPortalRoom
 oCPortalRoomManager
 oCSpawnManager
 oCTriggerChangeLevel
 oCTriggerScript
 oCViewStatusBar
 oCWay
 oCWorldTimer
+oCZoneMusic
+oSMenuInfoArmor
+oSMenuInfoAttribute
+oSMenuInfoTalent
+oTSpawnNode
 oWorld
 zCAIPlayer
 zCArray
 zCArraySort
+zCBspSector
 zCCamera
+zCClassDef
+zCConsole
 zCList
 zCListSort
+zCMaterial
 zCMenu
 zCMenuItem
+zCMenuItemText
+zCMesh
 zCMover
 zCObject
 zCOption
 zCOptionEntry
 zCOptionSection
 zCPar_Symbol
 zCParser
+zCPolygon 
+zCResource
+zCRoute
 zCSession
 zCSkyController_Outdoor
+zCTex_D3D
+zCTexture
 zCTimer
 zCTree
 zCTrigger
+zCVertFeature
 zCView
 zCViewText
+zCVisual
 zCVob
 zCVobLight
 zCVobWaypoint
 zCWayNet
 zCWaypoint
 zCZoneZFog
 zERROR
 zList
 zMAT4
 zMATRIX4
 zString
+zTBBox3D
+zTMov_Keyframe
+zTPlane 
+zTPortalInfo
```

I want to add the missing classes and then go over all of them to confirm the correct layout for both Gothic 1 and Gothic 2. I noticed on many occasions that the class definitions for Gothic 1 are just copies of Gothic 2 with incorrect layout.

That might take a while.
@Lehona Any ideas on how to do this efficiently?

**Comment by szapp** [comment]
These are the class sizes/checksums that I could find so far by using `zCClassDef.classSize` and `zCClassDef.archiveVersionSum` (green same, red different). Unfortunately, not all of the classes have a class definition (and/or a checksum for some reason even after calculating it with `zCClassDef::EndStartup(void)`).

I wonder if there is another way to obtain more information about the class variables (e.g. variable type in order of the memory layout).

```diff
 --------------------------------------------------------------------------
                                   SIZE                      CHECKSUM
                           Gothic 1    Gothic 2        Gothic 1    Gothic 2
 --------------------------------------------------------------------------
 CGameManager:              
+oCAniCtrl_Human:                  4600                                    
 oCBarrier:                 
 oCGame:                    
 oCInfo:                    
 oCInfoChoice:              
 oCInfoManager:             
 oCInformationManager :     
-oCItem:                     764         840           0162D780    0153D000
 oCLogManager:              
 oCLogTopic:                
 oCMag_Book:                
 oCMob:                     
-oCMobContainer:             608         644                       0082E6F0
-oCMobDoor:                  592         628                       0082E6F0
-oCMobFire:                  572         608                       0082E6F0
-oCMobInter:                 528         564                               
-oCMobLockable:              572         608           007D06E8    0082E6F0
 oCNews:                    
-oCNpc:                      2616        2564          4AB71B00    4AB71B00
+oCNpcTalent:                       48                 FFFFFFFF    FFFFFFFF
 oCPortalRoom:              
 oCPortalRoomManager:       
 oCSpawnManager:            
-oCTriggerChangeLevel:       368         400           FFFFFFFF    0082E6F0
-oCTriggerScript:            348         380           4AB71B00    4AB71B00
 oCViewStatusBar:           
 oCWay:                     
 oCWorldTimer:              
-oCZoneMusic:                296         328                               
 oSMenuInfoArmor:           
 oSMenuInfoAttribute:       
 oSMenuInfoTalent:          
 oTSpawnNode:               
 oWorld:                    
+zCAIPlayer:                       280                                     
 zCArray:                   
 zCArraySort:               
 zCBspSector:               
 zCCamera:                  
 zCClassDef:                
 zCConsole:                 
 zCList:                    
 zCListSort:                
-zCMaterial:                 148         172                               
 zCMenu:                    
 zCMenuItem:                
 zCMenuItemText:            
-zCMesh:                     224         228           007D06E8            
-zCMover:                    592         624                               
+zCObject:                         36                                      
 zCOption:                  
 zCOptionEntry:             
 zCOptionSection:           
 zCPar_Symbol:              
 zCParser:                  
 zCPolygon :                
+zCResource:                       84                                      
 zCRoute:                   
 zCSession:                 
 zCSkyController_Outdoor:   
 zCTex_D3D:                 
+zCTexture:                       140                             11E1A300
 zCTimer:                   
 zCTree:                    
-zCTrigger:                  328         360                               
 zCVertFeature:             
 zCView:                    
 zCViewText:                
+zCVisual:                         52                                      
-zCVob:                      256         288                               
-zCVobLight:                 348         392           0AF6A460    01558DD8
-zCVobWaypoint:              256         288                               
+zCWayNet:                         76                                      
+zCWaypoint:                      124                                     
-zCZoneZFog:                 272         312                               
 zERROR:                    
 zList:                     
 zMAT4:                     
 zMATRIX4:                  
 zString:                   
 zTBBox3D:                  
 zTMov_Keyframe:            
 zTPlane :                  
 zTPortalInfo:              
 --------------------------------------------------------------------------
```

**Comment by Lehona** [comment]
Some classes implement getter and/or setter methods, which are probably simple enough to even parse them automatically (and thus compare offsets across versions). Even with only a couple of those methods we could gain a high amount of certainty about similar memory layout.

There will be cascading effects as soon as one of the base classes changes its memory layout, though.

**Comment by szapp** [comment]
Thanks for your thoughts!

Yesterday I was mostly brute-forcing it, by manually comparing class sizes between Ikarus and what I could find in IDA in functions allocating memory before calling respective constructors. When the sizes matched for Gothic 2, I assumed the Ikarus classes to be correct (I don't doubt Sektenspinner and friends' research when it comes to Gothic 2).

As for Gothic 1, I plan to copy the missing classes from Gothic 2 if I find the class sizes to be identical in IDA. If not, I will follow your advice with getter and setter methods as well as looking into archiver functions (if they exist for that class).

Yesterday I was about half-way done with checking the class sizes. I will continue when I find the time.

**Comment by szapp** [comment]
This here is now a full list of all desired changes. I opened a PR #21 and will gradually (working this list from top to bottom) add commits with the required changes/additions.

## Class Size Mismatches

These are the classes that have different sizes in Ikarus compared to what I found in IDA. All other non-listed classes match in size.


&#9989; | Class | Size in Ikarus | Size in Engine
------- | ----- | ------------- | --------------
&nbsp; | **Gothic 1** | | 
f27e750  | zCSkyController_Outdoor |1664 | 1672
| | |
&nbsp; | **Gothic 2** | |
fe8e029  | oCNpcTalent | 12 | 48
52b58eb | zCClassDef | 112 | 108
9827f87 | zCRoute | 24 | 28

## Missing Classes in Gothic 1 Ikarus
These would be nice to add to Ikarus. Entries marked with * denote differing class size from Gothic 2 Ikarus. All the others could potentially just be copied.

&#9989; | Class | Size in Engine | Size in Ikarus (G2)
------- | ----- | -------------- | ------
e6f5c3c | oCBarrier           | 292  | **300** *
935069d | oCLogManager        | 8    | 8
935069d | oCLogTopic          | 36   | 36
d6bd32c | oCNpcTalent         | 48   | **12** *
df4bb1e | oCZoneMusic         | 296  | **328** *
4c87f12 | oSMenuInfoArmor     | 4    | 4
4c87f12 | oSMenuInfoAttribute | 52   | 52
4c87f12 | oSMenuInfoTalent    | 68   | 68
9d26b44 | oTSpawnNode         | ?    | 20
79d6deb  | zCBspSector         | ?    | 100
84280f3 | zCConsole           | 224  | **220** *
ba28896 | zCClassDef | 112 | **108** *
f119aac  | zCMaterial          | 148  | **172** *
b36c7bb | zCMenuItemText      | ?    | 1076
de11587 | zCMesh              | 224  | **228** *
7a7d9da | zCPolygon           | 56   | **52** *
02e08e3 | zCResource          | 84   | 84
7679de1  | zCRoute             | 28   | **24** *
9541e57 | zCTex_D3D           | 316  | 316
b76ee80 | zCTexture           | 140  | 140
d026255 | zCVertFeature       | 32   | **28** *
37ada73 | zCVisual            | 52   | 52
ff3f0f0 | zTBBox3D            | 24   | 24
056fcda | zTMov_Keyframe      | 28   | 28
e972699 | zTPlane             | 16   | 16
e85df53  | zTPortalInfo        | ?    | 8

## Missing Classes in Gothic 2 Ikarus
Analogous to the list above.

&#9989; | Class | Size in Engine | Size in Ikarus (G1)
------- | ----- | -------------- | ------
d1f6ba1 | oCNews | 100 | 100

**Comment by Lehona** [comment]
Closed via #21
